### ISSUE TITLE
Redundant separate implementations for const and non-const in ForEachIdx

### ISSUE THREAD

**Issue by sbastrakov** [ComputationalRadiationPhysics/picongpu#2877]
It seems the attempt to separate implementations for const and non-const functors, that currently are just copies of one another, are actually not needed. A simple way seems to be just using a universal reference and a single implementation, [a small example](https://coliru.stacked-crooked.com/a/08fdccb610283a68).

```C++
#include <iostream>


struct Functor {
   void operator()() const { std::cout << "const version\n"; }
   void operator()() { std::cout << "non-const version\n"; }
};

template< class T_Functor >
void apply( T_Functor && functor ) {
   functor();   
}


int main()
{
    const Functor constFunctor;
    Functor nonconstFunctor;
    apply(constFunctor);
    apply(nonconstFunctor);
    return 0;
}
```

Discussed offline with @psychocoderHPC . Also discussed on this class:

- theoretically we probably do not want a non-const version at all (since functors are generally better stateless), but currently need both since some `const`s are missing in other places;
- for (a rather distant) future we might want to use `std::invoke` there instead of the function call syntax, [to be more compatible with how C++17 treats Callables](http://ericniebler.com/2018/12/05/standard-ranges/); ofc it is for distant future and needs to be checked with CUDA, so really not the priority.

**Comment by sbastrakov** [comment]
Forgotten originally. Also discussed with @psychocoderHPC that it might make sense to split the main loop into two: one for a guaranteed multiple of SIMD width, and the trailing one. Then we could eliminate an if statement from the first loop. So basically how the compiler approaches vectorization of loops with a general number of iterations. (Maybe in our case the compiler could figure that itself, we are not sure).

**Comment by psychocoderHPC** [comment]
@sbastrakov also discussed that it could be that splitting the loops into two can increase the register footprint. This must be tested but we need to look into it!

**Comment by sbastrakov** [comment]
Closed with #2881 .
